cpp/src/replace/clamp.cu

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/replace.hpp>
+#include <cudf/detail/iterator.cuh>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+template <typename Transformer>
+std::pair<std::unique_ptr<column>, std::unique_ptr<column>>
+form_offsets_and_char_column (cudf::column_device_view input,
+                              Transformer offsets_transformer,
+                              rmm::mr::device_memory_resource* mr,
+                              cudaStream_t stream) {
+
+    std::unique_ptr<column> offsets_column{};
+    auto strings_count = input.size();
+
+    if (input.nullable()) {
+        auto input_begin = cudf::experimental::detail::make_null_replacement_iterator<string_view>(input, string_view{});
+        auto offsets_transformer_itr = thrust::make_transform_iterator(input_begin, offsets_transformer);
+        offsets_column = std::move(cudf::strings::detail::make_offsets_child_column(offsets_transformer_itr,
+                    offsets_transformer_itr + strings_count,
+                    mr, stream));
+    } else {
+        auto offsets_transformer_itr = thrust::make_transform_iterator(input.begin<string_view>(), offsets_transformer);
+        offsets_column = std::move(cudf::strings::detail::make_offsets_child_column(offsets_transformer_itr,
+                    offsets_transformer_itr + strings_count,
+                    mr, stream));
+    }
+
+    auto d_offsets = offsets_column->view().template data<size_type>();
+    // build chars column
+    size_type bytes = thrust::device_pointer_cast(d_offsets)[strings_count];
+    auto chars_column = cudf::strings::detail::create_chars_child_column( strings_count, input.null_count(), bytes, mr, stream);
+
+    return std::make_pair(std::move(offsets_column), std::move(chars_column));
+}
+
+template <typename ScalarIterator>
+std::unique_ptr<cudf::column> clamp_string_column (strings_column_view const& input,
```

Comment:
This function has a lot of repetition of identical code within the if/else if/else sections. Can it be factored out to reduce the duplication?